### PR TITLE
Add cli command to create MIR certificates

### PR DESCRIPTION
**Pull request by Jimbo4350**
- Move governance related code to `Cardano.CLI.Shelley.Run.Governance`

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 1]
```diff
@@ -442,7 +442,10 @@ pGovernanceCmd :: Parser GovernanceCmd
 pGovernanceCmd =
   Opt.subparser $
     mconcat
-      [ Opt.command "create-update-proposal"
+      [ Opt.command "create-MIR-certificate"
+          (Opt.info pMIRCertificate $
+            Opt.progDesc "Create an MIR certificate")
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:447]: Perhaps we should expand on the "MIR" abbreviation:

```suggestion
            Opt.progDesc "Create an MIR (Move Instantaneous Rewards) certificate")
```

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+
+import           Cardano.Config.Shelley.ColdKeys (KeyError, KeyRole(..), readVerKey,
+                   renderKeyError)
+import           Cardano.CLI.Shelley.Parsers
+
+data ShelleyGovernanceError
+  = GovernanceWriteMIRCertError !FilePath !ApiError
+  | GovernanceWriteUpdateError !FilePath !ApiError
+  | GovernanceEmptyUpdateProposal
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs:29]: Not really a part of this PR, but:

```suggestion
  | GovernanceEmptyUpdateProposalError
```

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    firstExceptT (GovernanceWriteUpdateError upFile)  . newExceptT $ writeUpdate upFile upProp
+  where
+    checkForEmptyProposal :: ShelleyPParamsUpdate -> ExceptT ShelleyGovernanceError IO ShelleyPParamsUpdate
+    checkForEmptyProposal sPParams
+      | sPParams == emptyPParamsUpdate = left GovernanceEmptyUpdateProposal
+      | otherwise = right sPParams
+
+-- TODO: Import from shelley ledger specs
+emptyPParamsUpdate :: ShelleyPParamsUpdate
+emptyPParamsUpdate =
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs:103]: I'll raise a PR in `shelley-ledger-specs` to export `emptyPParamsUpdate` so we can use it (I'll address this in a follow-up PR).

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs [hunk 4]
```diff
@@ -442,7 +442,10 @@ pGovernanceCmd :: Parser GovernanceCmd
 pGovernanceCmd =
   Opt.subparser $
     mconcat
-      [ Opt.command "create-update-proposal"
+      [ Opt.command "create-MIR-certificate"
```
Comment (intricate) [inline cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:445]: I need to check if this is case-sensitive. If so, I think we should instead use:

```suggestion
      [ Opt.command "create-mir-certificate"
```

Comment (intricate) [reply cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:445]: It appears this is case-sensitive. Not sure what others think, but I'd prefer sticking with commands and options in all lowercase.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs:103]: This needs to be double checked at the equality isn't working.

File: cardano-cli/src/Cardano/CLI/Shelley/Run.hs [hunk 5]
```diff
@@ -86,28 +86,57 @@ runBlockCmd cmd = liftIO $ putStrLn $ "TODO: runBlockCmd: " ++ show cmd
...
+  :: [VerificationKeyFile]
+  -- ^ Stake verification keys
+  -> [ShelleyCoin]
+  -- ^ Reward amounts
+  -> OutputFile
+  -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceMIRCertificate vKeys rwdAmts (OutputFile oFp) = do
+    sCreds <- mapM readStakeKey vKeys
+
+    let mirCert = shelleyMIRCertificate $ mirMap sCreds rwdAmts
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Shelley/Run.hs]: We should check that the number of `vKeys` and `rwdAmts` are the same and fail with an appropriate error if not. Currently we use `zip` below which means we ignore the longer ones from either side.

File: cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs [hunk 6]
```diff
@@ -93,7 +93,7 @@ runStakeKeyRegistrationCert (VerificationKeyFile vkFp) (OutputFile oFp) = do
 
 runStakeKeyDelegationCert
   :: VerificationKeyFile
-  -- ^ Delegator staking verification key file.
+  -- ^ Delegator stake verification key file.
```
Comment (dcoutts) [inline cardano-cli/src/Cardano/CLI/Shelley/Run/StakeAddress.hs:96]: :grin:

File: cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs (hunk 1 above)
Comment (dcoutts) [reply cardano-cli/src/Cardano/CLI/Shelley/Parsers.hs:447]: might as well yes

File: cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs (hunk 3 above)
Comment (intricate) [reply cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs:125]: This has been exported: https://github.com/input-output-hk/cardano-ledger-specs/pull/1498

**Pull request merged**